### PR TITLE
Add zio-telemetry opentracing backend

### LINKED ISSUE #477: Make ZIO instance work for MonadAsyncError[RIO[R, *]] and client aswell
- Make ZIO instance work for `MonadAsyncError[RIO[R, *]]` for task you can easily make it work with `MonadAsyncError[RIO[Any, *]]`
- Make the client also able to work with `RIO[R, *]` such that tracing implementations can be written.

I've made a sttp 1.x version (as I am stuck for the time being by a downward dependency)

```scala
import com.softwaremill.sttp.MonadAsyncError
import zio.{RIO, Task, UIO, ZIO}

class ZioMonadAsyncError[R] extends MonadAsyncError[RIO[R, *]] {
  override def unit[T](t: T): RIO[R, T] = ZIO.succeed(t)

  override def map[T, T2](fa: RIO[R, T])(f: T => T2): RIO[R, T2] = fa.map(f)

  override def flatMap[T, T2](fa: RIO[R, T])(f: T => RIO[R, T2]): RIO[R, T2] =
    fa.flatMap(f)

  override def async[T](register: (Either[Throwable, T] => Unit) => Unit): RIO[R, T] =
    ZIO.effectAsyncInterrupt { cb =>
      val canceler = register {
        case Left(t)  => cb(Task.fail(t))
        case Right(t) => cb(Task.succeed(t))
      }

      Left(UIO(canceler))
    }

  override def error[T](t: Throwable): RIO[R, T] = ZIO.fail(t)

  protected override def handleWrappedError[T](rt: RIO[R, T])(h: PartialFunction[Throwable, RIO[R, T]]): RIO[R, T] =
    rt.catchSome(h)

}
```

```scala
import io.netty.buffer.ByteBuf
import java.nio.ByteBuffer
import org.reactivestreams.Publisher
import org.asynchttpclient.proxy.ProxyServer
import org.asynchttpclient.{AsyncHttpClient, DefaultAsyncHttpClient, DefaultAsyncHttpClientConfig, Realm}
import com.softwaremill.sttp.{FollowRedirectsBackend, SttpBackend, SttpBackendOptions}
import com.softwaremill.sttp.SttpBackendOptions.ProxyType.{Http, Socks}
import com.softwaremill.sttp.asynchttpclient.AsyncHttpClientBackend
import zio._
import scala.jdk.CollectionConverters._
import com.github.ghik.silencer.silent

// Works with MonadAsyncError[RIO[R, *]] and sttp 1.x
object AsyncHttpClientZioBackend {

  @silent // dead code following this construct ??
  private class Backend[R](asyncHttpClient: AsyncHttpClient, closeClient: Boolean)
      extends AsyncHttpClientBackend[RIO[R, *], Nothing](asyncHttpClient, new ZioMonadAsyncError[R](), closeClient) {

    protected override def streamBodyToPublisher(s: Nothing): Publisher[ByteBuf] =
      s // nothing is everything

    protected override def publisherToStreamBody(p: Publisher[ByteBuffer]): Nothing =
      throw new IllegalStateException("This backend does not support streaming")

    protected override def publisherToBytes(p: Publisher[java.nio.ByteBuffer]): ZIO[R, Throwable, Array[Byte]] =
      throw new IllegalStateException("This backend does not support streaming")
  }

  type SttpClient[R] = Has[SttpBackend[RIO[R, *], Nothing]]

  private def withFollowRedirects[R](
    asyncHttpClient: AsyncHttpClient,
    closeClient: Boolean
  ): SttpBackend[RIO[R, *], Nothing] =
    new FollowRedirectsBackend[RIO[R, *], Nothing](
      new Backend[R](asyncHttpClient, closeClient)
    )

  private def create[R](
    options: SttpBackendOptions
  ): Task[SttpBackend[RIO[R, *], Nothing]] =
    Task.effect(
      withFollowRedirects[R](new DefaultAsyncHttpClient(defaultConfigBuilder(options).build()), closeClient = true)
    )

  def managed[R](
    options: SttpBackendOptions
  ): TaskManaged[SttpBackend[RIO[R, *], Nothing]] =
    ZManaged.make(create[R](options))(b => ZIO(b.close()).ignore)

  def layer[R: Tagged](
    options: SttpBackendOptions = SttpBackendOptions.Default
  ): Layer[Throwable, SttpClient[R]] =
    ZLayer.fromManaged(managed[R](options))

  private def defaultConfigBuilder(
    options: SttpBackendOptions
  ): DefaultAsyncHttpClientConfig.Builder = {
    val configBuilder = new DefaultAsyncHttpClientConfig.Builder()
      .setConnectTimeout(options.connectionTimeout.toMillis.toInt)

    options.proxy match {
      case None => configBuilder
      case Some(p) =>
        val proxyType: org.asynchttpclient.proxy.ProxyType =
          p.proxyType match {
            case Socks => org.asynchttpclient.proxy.ProxyType.SOCKS_V5
            case Http  => org.asynchttpclient.proxy.ProxyType.HTTP
          }

        configBuilder.setProxyServer {
          val builder = new ProxyServer.Builder(p.host, p.port)
            .setProxyType(proxyType) // Fix issue #145
            .setNonProxyHosts(p.nonProxyHosts.asJava)

          p.auth.foreach { proxyAuth =>
            builder.setRealm(
              new Realm.Builder(proxyAuth.username, proxyAuth.password).setScheme(Realm.AuthScheme.BASIC)
            )
          }

          builder.build()
        }
    }
  }

}

```

### PR DESCRIPTION
**Pull request by Fristi**
I modified the `TaskMonadAsyncError` to work with a `RIO[R, *]` so we can use a tracing effect in the zio-telemetry backend. This closes #477

### CODE REVIEW

**Comment by Fristi** [comment]
Build fails due

```
[ERROR] [04/24/2020 09:31:58.732] [sttp-test-server-akka.actor.default-dispatcher-14] [akka.actor.ActorSystemImpl(sttp-test-server)] Response stream for [GET /error] failed with 'null'. Aborting connection.

java.lang.RuntimeException

	at sttp.client.testing.server.HttpServer.$anonfun$serverRoutes$106(HttpServer.scala:284)

```

Or am I missing anything?

File: metrics/zio-telemetry-open-tracing-backend/src/main/scala/sttp/client/ziotelemetry/opentracing/ZioTelemetryOpenTracingBackend.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import sttp.client.monad.MonadError
+import sttp.client.ws.WebSocketResponse
+import zio._
+import zio.telemetry.opentracing._
+
+class ZioTelemetryOpenTracingBackend[-WS_HANLDER[_]](other: SttpBackend[Task, Nothing, WS_HANLDER], requestTagCollector: RequestTagCollector = RequestTagCollector.empty) extends SttpBackend[RIO[OpenTracing, *], Nothing, WS_HANLDER] {
+
+  @SuppressWarnings(Array("scalafix:Disable.toString"))
+  def send[T](request: Request[T, Nothing]): RIO[OpenTracing, Response[T]] = {
+    val headers = scala.collection.mutable.Map.empty[String, String]
```
Comment (adamw) [inline metrics/zio-telemetry-open-tracing-backend/src/main/scala/sttp/client/ziotelemetry/opentracing/ZioTelemetryOpenTracingBackend.scala:17]: so the mutable `headers` map is to pass the injected opentracing headers to the request?

**Comment by adamw** [comment]
Looks good, thank you!

One missing piece, maybe you could add some docs? That would be probably somewhere in `docs/backends/wrappers`, there's a couple of others which might be an inspiration :)

File: metrics/zio-telemetry-open-tracing-backend/src/main/scala/sttp/client/ziotelemetry/opentracing/ZioTelemetryOpenTracingBackend.scala (hunk 1 above)
Comment (Fristi) [reply metrics/zio-telemetry-open-tracing-backend/src/main/scala/sttp/client/ziotelemetry/opentracing/ZioTelemetryOpenTracingBackend.scala:17]: Yes it creates from the existing Span context a set of headers which are put on the request such that distributed tracing is enabled

**Comment by Fristi** [comment]
Added some docs

**Comment by adamw** [comment]
Released in 2.1.0, thanks!

**Pull request merged**